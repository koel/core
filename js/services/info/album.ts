import { secondsToHis } from '@/utils'
import { http } from '..'

export const albumInfo = {
  fetch (album: Album): Promise<Album> {
    return new Promise((resolve, reject): void => {
      if (album.info) {
        resolve(album)
        return
      }

      http.get(`album/${album.id}/info`, ({ data }: { data: AlbumInfo | null }): void => {
        data && this.merge(album, data)
        resolve(album)
      }, (error: any) => reject(error))
    })
  },

  /**
   * Merge the (fetched) info into an album.
   */
  merge: (album: Album, info: AlbumInfo): void => {
    // Convert the duration into i:s
    info.tracks && info.tracks.forEach(track => (track.fmtLength = secondsToHis(track.length)))

    // If the album cover is not in a nice form, discard.
    if (typeof info.image !== 'string') {
      info.image = null
    }

    // Set the album cover on the client side to the retrieved image from server.
    if (info.image) {
      album.cover = info.image
    }

    album.info = info
  }
}
