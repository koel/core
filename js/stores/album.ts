/* eslint camelcase: ["error", {properties: "never"}] */
import Vue from 'vue'
import { union, difference, take, orderBy } from 'lodash'

import stub from '@/stubs/album'
import { artistStore } from '.'
import { http } from '@/services'
import { use } from '@/utils'

export const albumStore = {
  stub,
  cache: {} as { [key: string]: Album },

  state: {
    albums: [stub]
  },

  init (albums: Album[]) {
    // Traverse through the artists array and add their albums into our master album list.
    this.all = albums
    this.all.forEach(album => this.setupAlbum(album))
  },

  setupAlbum (album: Album): void {
    const artist = artistStore.byId(album.artist_id)!
    artist.albums = union(artist.albums, [album])

    Vue.set(album, 'artist', artist)
    Vue.set(album, 'info', null)
    Vue.set(album, 'songs', [])
    Vue.set(album, 'playCount', 0)

    this.cache[album.id] = album
  },

  get all () {
    return this.state.albums
  },

  set all (value) {
    this.state.albums = value
  },

  byId (id: number): Album | undefined {
    return this.cache[id]
  },

  byIds (ids: number[]): Album[] {
    const albums = [] as Album[]
    ids.forEach(id => use(this.byId(id), album => albums.push(album!)))
    return albums
  },

  add (albums: Album | Album[]): void {
    (<Album[]>[]).concat(albums).forEach(album => {
      this.setupAlbum(album)
      album.playCount = album.songs.reduce((count, song) => count + song.playCount, 0)
      this.all.push(album)
    })
  },

  purify (): void {
    this.compact()
  },

  /**
   * Remove empty albums from the store.
   */
  compact (): void {
    const emptyAlbums = this.all.filter(album => album.songs.length === 0)
    if (!emptyAlbums.length) {
      return
    }

    this.all = difference(this.all, emptyAlbums)
    emptyAlbums.forEach(album => delete this.cache[album.id])
  },

  getMostPlayed (n: number = 6): Album[] {
    // Only non-unknown albums with actual play count are applicable.
    const applicable = this.all.filter(album => album.playCount && album.id !== 1)
    return take(orderBy(applicable, 'playCount', 'desc'), n)
  },

  getRecentlyAdded (n: number = 6): Album[] {
    const applicable = this.all.filter(album => album.id !== 1)
    return take(orderBy(applicable, 'created_at', 'desc'), n)
  },

  /**
   * Upload a cover for an album.
   *
   * @param {Album} album The album object
   * @param {string} cover The content data string of the cover
   */
  uploadCover: async (album: Album, cover: string): Promise<string> => {
    const { coverUrl } = await http.put<{ coverUrl: string }>(`album/${album.id}/cover`, { cover })
    album.cover = coverUrl
    return album.cover
  },

  /**
   * Get the (blurry) thumbnail-sized version of an album's cover.
   */
  getThumbnail: async (album: Album): Promise<string | null> => {
    if (album.thumbnail === undefined) {
      const { thumbnailUrl } = await http.get<{ thumbnailUrl: string }>(`album/${album.id}/thumbnail`)

      album.thumbnail = thumbnailUrl
    }

    return album.thumbnail
  }
}
