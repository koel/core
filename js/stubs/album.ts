import artist from './artist'

const albumInfo: AlbumInfo = {
  image: '',
  tracks: []
}

const album: Album = {
  artist,
  id: 0,
  artist_id: 0,
  name: '',
  cover: '',
  playCount: 0,
  length: 0,
  fmtLength: '00:00',
  songs: [],
  is_compilation: false,
  info: albumInfo
}

export default album
