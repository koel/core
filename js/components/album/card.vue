<template>
  <article
    :title="`${album.name} by ${album.artist.name}`"
    @dragstart="dragStart"
    class="item"
    draggable="true"
    tabindex="0"
    v-if="album.songs.length"
  >
    <span class="thumbnail-wrapper">
      <album-thumbnail :entity="album" />
    </span>

    <footer>
      <div class="info">
        <a class="name" :href="`#!/album/${album.id}`">{{ album.name }}</a>
        <span class="sep">by</span>
        <a
          class="artist"
          v-if="isNormalArtist"
          :href="`#!/artist/${album.artist.id}`"
        >{{ album.artist.name }}</a>
        <span class="artist nope" v-else>{{ album.artist.name }}</span>
      </div>
      <p class="meta">
        <span class="left">
          {{ album.songs.length | pluralize('song') }}
          •
          {{ fmtLength }}
          •
          {{ album.playCount | pluralize('play') }}
        </span>
        <span class="right">
          <a
            :title="`Shuffle all songs in the album ${album.name}`"
            @click.prevent="shuffle"
            class="shuffle-album"
            href
            role="button"
          >
            <i class="fa fa-random"></i>
          </a>
          <a
            :title="`Download all songs in the album ${album.name}`"
            @click.prevent="download"
            class="download-album"
            href
            role="button"
            v-if="sharedState.allowDownload"
          >
            <i class="fa fa-download"></i>
          </a>
        </span>
      </p>
    </footer>
  </article>
</template>

<script lang="ts">
import mixins from 'vue-typed-mixins'
import { pluralize, startDragging } from '@/utils'
import { artistStore, sharedStore } from '@/stores'
import { playback, download } from '@/services'
import albumAttributes from '@/mixins/album-attributes.ts'

export default mixins(albumAttributes).extend({
  components: {
    AlbumThumbnail: () => import('@/components/ui/album-artist-thumbnail.vue')
  },

  filters: { pluralize },

  data: () => ({
    sharedState: sharedStore.state
  }),

  computed: {
    isNormalArtist (): boolean {
      return !artistStore.isVariousArtists(this.album.artist) &&
        !artistStore.isUnknownArtist(this.album.artist)
    }
  },

  methods: {
    shuffle (): void {
      playback.playAllInAlbum(this.album, true /* shuffled */)
    },

    download (): void {
      download.fromAlbum(this.album)
    },

    dragStart (event: DragEvent): void {
      startDragging(event, this.album, 'Album')
    }
  }
})
</script>

<style lang="scss">
@import "~#/partials/_vars.scss";
@import "~#/partials/_mixins.scss";

.sep {
  display: none;
  color: $color2ndText;

  .as-list & {
    display: inline;
  }
}

@include artist-album-card();
</style>
