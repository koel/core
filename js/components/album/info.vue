<template>
  <article class="album-info" :class="mode">
    <h1 class="name">
      <span>{{ album.name }}</span>
      <a
        :title="`Shuffle all songs in ${album.name}`"
        @click.prevent="shuffleAll"
        class="shuffle"
        role="button"
      >
        <i class="fa fa-random"></i>
      </a>
    </h1>

    <main>
      <album-thumbnail :entity="album"/>

      <template v-if="album.info">
        <div class="wiki" v-if="album.info.wiki && album.info.wiki.summary">
          <div class="summary" v-show="showSummary" v-html="album.info.wiki.summary"></div>
          <div class="full" v-show="showFull" v-html="album.info.wiki.full"></div>

          <button class="more" v-if="showSummary" @click.prevent="showingFullWiki = true">
            Full Wiki
          </button>
        </div>

        <track-list :album="album" v-if="album.info.tracks && album.info.tracks.length"/>

        <footer>Data &copy; <a target="_blank" :href="album.info.url">Last.fm</a></footer>
      </template>
    </main>

  </article>
</template>

<script lang="ts">
import { sharedStore } from '@/stores'
import { playback, ls } from '@/services'
import Vue, { PropOptions } from 'vue'

export default Vue.extend({
  props: {
    album: Object as PropOptions<Album>,
    mode: {
      type: String,
      default: 'sidebar',
      validator: value => ['sidebar', 'full'].includes(value)
    }
  },

  components: {
    TrackList: () => import('./track-list.vue'),
    AlbumThumbnail: () => import('@/components/ui/album-artist-thumbnail.vue')
  },

  data: () => ({
    showingFullWiki: false,
    useiTunes: sharedStore.state.useiTunes
  }),

  watch: {
    /**
     * Whenever a new album is loaded into this component, we reset the "full wiki" state.
     */
    album (): void {
      this.showingFullWiki = false
    }
  },

  computed: {
    showSummary (): boolean {
      return this.mode !== 'full' && !this.showingFullWiki
    },

    showFull (): boolean {
      return this.mode === 'full' || this.showingFullWiki
    },

    iTunesUrl (): string {
      return `${window.BASE_URL}api/itunes/album/${this.album.id}&jwt-token=${ls.get('jwt-token')}`
    }
  },

  methods: {
    shuffleAll (): void {
      playback.playAllInAlbum(this.album)
    }
  }
})
</script>

<style lang="scss">
@import "~#/partials/_vars.scss";
@import "~#/partials/_mixins.scss";

.album-info {
  @include artist-album-info();
}
</style>
