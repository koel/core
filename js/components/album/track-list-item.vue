<template>
  <li :class="{ available: song }" :title="tooltip" @click="play" role="button" tabindex="0">
    <span class="no">{{ index + 1 }}</span>
    <span class="title">{{ track.title }}</span>
    <a
      :href="iTunesUrl"
      v-if="useiTunes && !song"
      target="_blank"
      class="view-on-itunes"
      title="View on iTunes"
      >
      iTunes
    </a>
    <span class="length">{{ track.fmtLength }}</span>
  </li>
</template>

<script lang="ts">
import Vue, { PropOptions } from 'vue'
import { songStore, queueStore, sharedStore } from '@/stores'
import { ls, playback } from '@/services'

export default Vue.extend({
  props: {
    album: {
      type: Object,
      required: true
    } as PropOptions<Album>,

    track: {
      type: Object,
      required: true
    } as PropOptions<AlbumTrack>,

    index: {
      type: Number,
      required: true
    }
  },

  data: () => ({
    useiTunes: sharedStore.state.useiTunes
  }),

  computed: {
    song (): Song | null {
      return songStore.guess(this.track.title, this.album)
    },

    tooltip (): string {
      return this.song ? 'Click to play' : ''
    },

    iTunesUrl (): string {
      return `${window.BASE_URL}api/itunes/song/${this.album.id}?q=${encodeURIComponent(this.track.title)}&jwt-token=${ls.get('jwt-token')}`
    }
  },

  methods: {
    play (): void {
      if (this.song) {
        queueStore.contains(this.song) || queueStore.queueAfterCurrent(this.song)
        playback.play(this.song)
      }
    }
  }
})
</script>

<style lang="scss" scoped>
@import "~#/partials/_vars.scss";

[role=button] {
  &:focus {
    span.title {
      color: $colorHighlight;
    }
  }

  a.view-on-itunes {
    display: inline-block;
    border-radius: 3px;
    font-size: .8rem;
    padding: 0 5px;
    color: #fff;
    background: rgba(255, 255, 255, .1);
    height: 20px;
    line-height: 20px;
    margin-left: 4px;

    &:hover, &:focus {
      background: linear-gradient(27deg, #fe5c52 0%,#c74bd5 50%,#2daaff 100%);
      color: #fff;
    }

    &:active {
      box-shadow: inset 0px 5px 5px -5px #000;
    }
  }
}
</style>
