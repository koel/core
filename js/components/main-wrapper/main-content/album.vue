<template>
  <section id="albumWrapper">
    <h1 class="heading">
      <span class="overview">
        <img :src="album.cover" width="64" height="64" class="cover">
        {{ album.name }}
        <controls-toggler :showing-controls="showingControls" @toggleControls="toggleControls"/>

        <span class="meta" v-show="album.songs.length">
          by
          <a class="artist" v-if="isNormalArtist" :href="`#!/artist/${album.artist.id}`">{{ album.artist.name }}</a>
          <span class="nope" v-else>{{ album.artist.name }}</span>
          •
          <template v-if="album.year > 0"><span>{{ album.year }}</span>•</template>
          {{ album.songs.length | pluralize('song') }}
          •
          {{ fmtLength }}

          <template v-if="sharedState.useLastfm">
            •
            <a class="info" href @click.prevent="showInfo" title="View album's extra information">Info</a>
          </template>
          <template v-if="sharedState.allowDownload">
            •
            <a class="download" href @click.prevent="download" title="Download all songs in album">
              Download All
            </a>
          </template>
        </span>
      </span>

      <song-list-controls
        v-show="album.songs.length && (!isPhone || showingControls)"
        @shuffleAll="shuffleAll"
        @shuffleSelected="shuffleSelected"
        :config="songListControlConfig"
        :selectedSongs="selectedSongs"
      />
    </h1>

    <song-list :items="album.songs" type="album" ref="songList"/>

    <section class="info-wrapper" v-if="sharedState.useLastfm && info.showing">
      <a href class="close" @click.prevent="info.showing = false"><i class="fa fa-times"></i></a>
      <div class="inner">
        <div class="loading" v-if="info.loading"><sound-bar/></div>
        <album-info :album="album" mode="full" v-else/>
      </div>
    </section>
  </section>
</template>

<script>
import { pluralize, event } from '@/utils'
import { albumStore, artistStore, sharedStore } from '@/stores'
import { playback, download, albumInfo as albumInfoService } from '@/services'
import router from '@/router'
import hasSongList from '@/mixins/has-song-list'
import albumAttributes from '@/mixins/album-attributes'
import albumInfo from '@/components/main-wrapper/extra/album-info.vue'
import soundBar from '@/components/shared/sound-bar.vue'

export default {
  name: 'main-wrapper--main-content--album',
  mixins: [hasSongList, albumAttributes],
  components: { albumInfo, soundBar },
  filters: { pluralize },

  data () {
    return {
      sharedState: sharedStore.state,
      album: albumStore.stub,
      info: {
        showing: false,
        loading: true
      }
    }
  },

  computed: {
    isNormalArtist () {
      return !artistStore.isVariousArtists(this.album.artist) &&
        !artistStore.isUnknownArtist(this.album.artist)
    }
  },

  watch: {
    /**
     * Watch the album's song count.
     * If this is changed to 0, the user has edit the songs in this album
     * and move all of them into another album.
     * We should then go back to the album list.
     */
    'album.songs.length': newSongCount => newSongCount || router.go('albums')
  },

  created () {
    /**
     * Listen to 'main-content-view:load' event to load the requested album
     * into view if applicable.
     *
     * @param {String} view   The view name
     * @param {Object} album  The album object
     */
    event.on(event.$names.LOAD_MAIN_CONTENT, (view, album) => {
      if (view === 'album') {
        this.info.showing = false
        this.album = album
        // #530
        this.$nextTick(() => this.$refs.songList && this.$refs.songList.sort())
      }
    })
  },

  methods: {
    /**
     * Shuffle the songs in the current album.
     * Overriding the mixin.
     */
    shuffleAll () {
      playback.queueAndPlay(this.album.songs, true /* shuffled */)
    },

    /**
     * Download all songs from the album.
     */
    download () {
      download.fromAlbum(this.album)
    },

    async showInfo () {
      this.info.showing = true
      if (!this.album.info) {
        this.info.loading = true
        try {
          await albumInfoService.fetch(this.album)
        } catch (e) {
        } finally {
          this.info.loading = false
        }
      } else {
        this.info.loading = false
      }
    }
  }
}
</script>

<style lang="scss" scoped>
@import "~#/partials/_vars.scss";
@import "~#/partials/_mixins.scss";

#albumWrapper {
  button.play-shuffle {
    i {
      margin-right: 0 !important;
    }
  }

  .heading {
    .overview {
      position: relative;
      padding-left: 84px;

      @media only screen and (max-width : 768px) {
        padding-left: 0;
      }
    }

    .cover {
      position: absolute;
      left: 0;

      @media only screen and (max-width : 768px) {
        display: none;
      }
    }

    a.artist {
      color: $colorMainText;
      display: inline;

      &:hover {
        color: $colorHighlight;
      }
    }
  }

  @include artist-album-info-wrapper();
}
</style>
