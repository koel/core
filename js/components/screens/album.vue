<template>
  <section id="albumWrapper">
    <screen-header>
      {{ album.name }}
      <controls-toggler :showing-controls="showingControls" @toggleControls="toggleControls"/>

      <template v-slot:thumbnail>
        <album-thumbnail :entity="album"/>
      </template>

      <template v-slot:meta>
        <span v-if="album.songs.length">
          by
          <a class="artist" v-if="isNormalArtist" :href="`#!/artist/${album.artist.id}`">{{ album.artist.name }}</a>
          <span class="nope" v-else>{{ album.artist.name }}</span>
          •
          {{ album.songs.length | pluralize('song') }}
          •
          {{ fmtLength }}

          <template v-if="sharedState.useLastfm">
            •
            <a class="info" href @click.prevent="showInfo" title="View album's extra information">Info</a>
          </template>
          <template v-if="sharedState.allowDownload">
            •
            <a class="download" href @click.prevent="download" title="Download all songs in album" role="button">
              Download All
            </a>
          </template>
        </span>
      </template>

      <template v-slot:controls>
        <song-list-controls
          v-if="album.songs.length && (!isPhone || showingControls)"
          @playAll="playAll"
          @playSelected="playSelected"
          :songs="album.songs"
          :config="songListControlConfig"
          :selectedSongs="selectedSongs"
        />
      </template>
    </screen-header>

    <song-list :items="album.songs" type="album" ref="songList"/>

    <section class="info-wrapper" v-if="sharedState.useLastfm && meta.showing">
      <close-modal-btn @click="meta.showing = false"/>
      <div class="inner">
        <div class="loading" v-if="meta.loading">
          <sound-bar/>
        </div>
        <album-info :album="album" mode="full" v-else/>
      </div>
    </section>
  </section>
</template>

<script lang="ts">
import mixins from 'vue-typed-mixins'
import { pluralize } from '@/utils'
import { artistStore, sharedStore } from '@/stores'
import { download, albumInfo as albumInfoService } from '@/services'
import router from '@/router'
import hasSongList from '@/mixins/has-song-list.ts'
import albumAttributes from '@/mixins/album-attributes.ts'

export default mixins(hasSongList, albumAttributes).extend({
  components: {
    ScreenHeader: () => import('@/components/ui/screen-header.vue'),
    AlbumInfo: () => import('@/components/album/info.vue'),
    SoundBar: () => import('@/components/ui/sound-bar.vue'),
    AlbumThumbnail: () => import('@/components/ui/album-artist-thumbnail.vue'),
    CloseModalBtn: () => import('@/components/ui/close-modal-btn.vue')
  },

  filters: { pluralize },

  data: () => ({
    sharedState: sharedStore.state,
    meta: {
      showing: false,
      loading: true
    }
  }),

  computed: {
    isNormalArtist (): boolean {
      return !artistStore.isVariousArtists(this.album.artist) &&
        !artistStore.isUnknownArtist(this.album.artist)
    }
  },

  watch: {
    /**
     * Watch the album's song count.
     * If this is changed to 0, the user has edit the songs in this album
     * and move all of them into another album.
     * We should then go back to the album list.
     */
    'album.songs.length': (newSongCount: number): void => {
      if (!newSongCount) {
        router.go('albums')
      }
    },

    album (): void {
      this.meta.showing = false
      // #530
      if (this.$refs.songList) {
        (this.$refs.songList as any).sort()
      }
    }
  },

  methods: {
    download (): void {
      download.fromAlbum(this.album)
    },

    async showInfo (): Promise<void> {
      this.meta.showing = true

      if (!this.album.info) {
        try {
          await albumInfoService.fetch(this.album)
        } catch (e) {
          /* eslint no-console: 0 */
          console.error(e)
        } finally {
          this.meta.loading = false
        }
      } else {
        this.meta.loading = false
      }
    }
  }
})
</script>

<style lang="scss" scoped>
@import "~#/partials/_vars.scss";
@import "~#/partials/_mixins.scss";

#albumWrapper {
  @include artist-album-info-wrapper();
}
</style>
