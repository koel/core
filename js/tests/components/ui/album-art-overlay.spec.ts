import each from 'jest-each'
import factory from '@/tests/factory'
import { albumStore } from '@/stores/album'
import Component from '@/components/ui/album-art-overlay.vue'
import { mount } from '@/tests/adapter'

describe('components/ui/album-art-overlay', () => {
  const cases = [
    ['no song playing', null],
    ['album having a cover same as stub', factory('album', {
      cover: albumStore.stub.cover
    })],
    ['album having a real cover', factory('album', {
      cover: '/img/foo.jpg'
    })]
  ]

  each(cases).test('renders property with %s', (description, album) => {
    const wrapper = mount(Component, { propsData: { album } })
    expect(wrapper).toMatchSnapshot()
  })
})
