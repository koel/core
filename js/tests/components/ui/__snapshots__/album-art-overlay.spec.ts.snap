// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/ui/album-art-overlay renders property with album having a cover same as stub 1`] = `<div style="background-image: none;"></div>`;

exports[`components/ui/album-art-overlay renders property with album having a real cover 1`] = `<div style="background-image: url(/img/foo.jpg);"></div>`;

exports[`components/ui/album-art-overlay renders property with no song playing 1`] = `<div style="background-image: none;"></div>`;
